### PR TITLE
Scrapper, Christina Nigamedzyanova - 21FPL2

### PR DESCRIPTION
**Pull request by christine-ni**

### CODE REVIEW

File: seminars/index.html [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
```
Comment (demid5111) [inline seminars/index.html]: remove any unrelated files

File: seminars/main.py
Comment (demid5111) [inline seminars/main.py:1]: remove unrelated files

**Review by demid5111 — CHANGES_REQUESTED** [review]
almost nothing to review. hurry up!

File: seminars/main.py
Comment (demid5111) [inline seminars/main.py:1]: discard unrelated changes

File: lab_5_scrapper/scrapper_config.json [hunk 2]
```diff
@@ -1,9 +1,10 @@
 {
-    "seed_urls": [],
-    "headers": {},
-    "total_articles_to_find_and_parse": 0,
-    "encoding": "",
-    "timeout": 0,
+    "seed_urls": ["https://www.vgoroden.ru/"],
+    "headers": {"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36(KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36'"
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json]: better layout

File: lab_5_scrapper/scrapper_config.json [hunk 3]
```diff
@@ -1,9 +1,10 @@
 {
-    "seed_urls": [],
-    "headers": {},
-    "total_articles_to_find_and_parse": 0,
-    "encoding": "",
-    "timeout": 0,
+    "seed_urls": ["https://www.vgoroden.ru/"],
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json]: will it be enough to collect 100 articles?

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -120,45 +229,73 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        date_bs = article_soup.find('time', {'class': 'toolbar__text'})['datetime']
+        date_and_time = ' '.join(re.findall(r'\d{4}-\d{2}-\d{2}', date_bs) + re.findall(r'\d{2}:\d{2}:\d{2}', date_bs))
+        date = self.unify_date_format(date_and_time)
 
     def unify_date_format(self, date_str: str) -> datetime.datetime:
         """
         Unifies date format
         """
-        pass
+        dt_object = datetime.datetime.strptime(date_str, "%Y-%m-%d %H:%M:%S")
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: unnecessary variable

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -120,45 +229,73 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        art_text = [i.text for i in article_text]
+        self.article.text = '\n'.join(art_text)
 
     def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> None:
         """
         Finds meta information of article
         """
-        pass
+        self.article.title = article_soup.find('h1', {'class': 'title'}).text
+        self.article.author = article_soup.find('span', {'class': 'toolbar-opposite__author-text'}).text
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: author should be a list

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -120,45 +229,73 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+        date_bs = article_soup.find('time', {'class': 'toolbar__text'})['datetime']
+        date_and_time = ' '.join(re.findall(r'\d{4}-\d{2}-\d{2}', date_bs) + re.findall(r'\d{2}:\d{2}:\d{2}', date_bs))
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: why do not you collect topics? I have easily found them in HTML of your website

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -90,25 +184,40 @@ def __init__(self, config: Config) -> None:
...
+            return 'https://www.vgoroden.ru' + article_bs['href']
+
+
 
     def find_articles(self) -> None:
         """
         Finds articles
         """
-        pass
+        #считывает только 36 ссылок, нужно исправить
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: 1. put as minimal number of comments as possible
2. forget russian when making comments in code (!)

File: lab_5_scrapper/scrapper.py [hunk 8]
```diff
@@ -90,25 +184,40 @@ def __init__(self, config: Config) -> None:
...
-        pass
+        self.config = config
+        self.urls = []
 
     def _extract_url(self, article_bs: BeautifulSoup) -> str:
         """
         Finds and retrieves URL from HTML
         """
-        pass
+        if re.fullmatch(r'/novosti/.+', article_bs['href']):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: how do you guarantee that there will be always a href attribute?

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -13,70 +56,121 @@ def __init__(self, path_to_config: Path) -> None:
...
 
 def make_request(url: str, config: Config) -> requests.models.Response:
     """
     Delivers a response from a request
     with given configuration
     """
-    pass
+    time.sleep((random.randint(3, 7)))
+    response = requests.get(url, timeout=config.get_timeout(), headers=config.get_headers())
+    return response
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:188]: set encoding to response before return

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -13,70 +56,121 @@ def __init__(self, path_to_config: Path) -> None:
...
+        if config.total_articles > 150:
+            raise NumberOfArticlesOutOfRangeError
+
+        if not isinstance(config.headers, dict):
+            raise IncorrectHeadersError
+
+        if not isinstance(config.encoding, str):
+            raise IncorrectEncodingError
+
+        if not isinstance(config.timeout, int) or config.timeout < 0 or config.timeout > 60:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: all constants should be taken from constants

File: lab_5_scrapper/scrapper.py [hunk 11]
```diff
@@ -13,70 +56,121 @@ def __init__(self, path_to_config: Path) -> None:
...
         are not corrupt
         """
-        pass
+        config = self._extract_config_content()
+
+        if not isinstance(config.seed_urls, list):
+            raise IncorrectSeedURLError
+
+        for url in config.seed_urls:
+            if not re.fullmatch(r'https://w?w?w?.+', url):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: do you think it is possible to have for example just one w?

File: lab_5_scrapper/scrapper.py [hunk 12]
```diff
@@ -13,70 +56,121 @@ def __init__(self, path_to_config: Path) -> None:
...
+        with open(self.path_to_config, 'r', encoding='utf-8') as f:
+            config_content = json.load(f)
+            seed_urls = config_content['seed_urls']
+            headers = config_content['headers']
+            num_of_articles = config_content['total_articles_to_find_and_parse']
+            encoding = config_content['encoding']
+            timeout = config_content['timeout']
+            should_verify_certificate = config_content['should_verify_certificate']
+            headless_mode = config_content['headless_mode']
+        return ConfigDTO(seed_urls, num_of_articles,  headers, encoding,
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: use dict unpacking for very short DTO initialization

File: lab_5_scrapper/scrapper.py [hunk 13]
```diff
@@ -1,8 +1,51 @@
...
+import requests
+from bs4 import BeautifulSoup
+
+from core_utils.article.article import Article
+from core_utils.article.io import to_raw, to_meta
+from core_utils.config_dto import ConfigDTO
+from core_utils.constants import ASSETS_PATH, CRAWLER_CONFIG_PATH, TIMEOUT_LOWER_LIMIT, TIMEOUT_UPPER_LIMIT
+
+
+class IncorrectSeedURLError(Exception):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:24]: provide meaningful docstrings to errors

**Review by demid5111 — CHANGES_REQUESTED** [review]
only 6?? you should make at least for 8

File: lab_5_scrapper/scrapper.py [hunk 14]
```diff
@@ -120,45 +255,94 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
 def main() -> None:
     """
     Entrypoint for scrapper module
     """
-    pass
+    configuration = Config(path_to_config=CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH)
+    crawler = Crawler(configuration)
+    crawler.find_articles()
+    print(crawler.get_search_urls())
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: remove unnecessary prints

File: lab_5_scrapper/scrapper.py [hunk 15]
```diff
@@ -120,45 +255,94 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
+    print(len(crawler.get_search_urls()))
+    for i, full_url in enumerate(crawler.urls, start=1):
+        parser = HTMLParser(full_url=full_url, article_id=i, config=configuration)
+        article = parser.parse()
+        if isinstance(article, Article):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:400]: do you in fact just that it is not None?

File: lab_5_scrapper/scrapper.py [hunk 16]
```diff
@@ -120,45 +255,94 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
 def prepare_environment(base_path: Union[Path, str]) -> None:
     """
     Creates ASSETS_PATH folder if no created and removes existing folder
     """
-    pass
+    if base_path.exists():
+        shutil.rmtree(base_path)
+    base_path.mkdir(parents=True)
+
+class CrawlerRecursive(Crawler):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:310]: no implementation so far...

File: lab_5_scrapper/scrapper.py [hunk 17]
```diff
@@ -120,45 +255,94 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
...
         """
         Finds meta information of article
         """
-        pass
+        title = article_soup.find('h1', {'class': 'title'}).text
+        if title:
+            self.article.title = title
+        author = article_soup.find('span', {'class': 'toolbar-opposite__author-text'}).text
+        if author:
+            self.article.author.append(author)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:273]: if no author it should be a list with NOT FOUND, remember?

File: lab_5_scrapper/scrapper.py [hunk 18]
```diff
@@ -90,25 +208,42 @@ def __init__(self, config: Config) -> None:
...
+        self.urls = []
+        self.seed_urls = self.config.get_seed_urls()
 
     def _extract_url(self, article_bs: BeautifulSoup) -> str:
         """
         Finds and retrieves URL from HTML
         """
-        pass
+        href = article_bs.get('href')
+        if isinstance(href, str) and re.fullmatch(r'/novosti/.+', href):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: again check for None instead of type

File: lab_5_scrapper/scrapper.py [hunk 19]
```diff
@@ -90,25 +208,42 @@ def __init__(self, config: Config) -> None:
...
         """
         Finds articles
         """
-        pass
+        for url in self.seed_urls:
+            response = make_request(url, self.config)
+            res_bs = BeautifulSoup(response.text, 'lxml')
+            for link in res_bs.find_all('a'):
+                article_url = self._extract_url(link)
+                if article_url is None or article_url == ' ':
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: check for duplicates

File: lab_5_scrapper/scrapper.py (hunk 9 above)
Comment (demid5111) [reply lab_5_scrapper/scrapper.py:188]: still actual

File: lab_5_scrapper/scrapper.py [hunk 20]
```diff
@@ -13,70 +77,124 @@ def __init__(self, path_to_config: Path) -> None:
...
+        if total_articles > NUM_ARTICLES_UPPER_LIMIT:
+            raise NumberOfArticlesOutOfRangeError
+
+        if not isinstance(headers, dict):
+            raise IncorrectHeadersError
+
+        if not isinstance(encoding, str):
+            raise IncorrectEncodingError
+
+        if not isinstance(timeout, int) or timeout < 0 or timeout > 60:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: constants should be taken from constants

File: lab_5_scrapper/scrapper.py [hunk 21]
```diff
@@ -13,70 +77,124 @@ def __init__(self, path_to_config: Path) -> None:
...
+
 
     def _extract_config_content(self) -> ConfigDTO:
         """
         Returns config values
         """
-        pass
+        with open(self.path_to_config, 'r', encoding='utf-8') as f:
+            config_content = json.load(f)
+        return ConfigDTO(config_content['seed_urls'],
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py]: use dictionary unpacking

**Review by demid5111 — CHANGES_REQUESTED** [review]
very nice, continue

File: lab_5_scrapper/scrapper.py (hunk 15 above)
Comment (christine-ni) [reply lab_5_scrapper/scrapper.py:400]: In case i do not check if the article value belongs to Article class, i have mistakes in mypy, and this is the only way i've found to pass the checks

**Comment by demid5111** [comment]
nothing to review to this moment

**Comment by vitazaytseva** [comment]
Well done! You can proceed to work on Lab 6)
